go/vt/wrangler/stream_migrater.go

```diff
@@ -0,0 +1,604 @@
+	if len(tmpl) == 0 {
+		return nil
+	}
+	return sm.mi.forAllTargets(func(target *miTarget) error {
+		tabletStreams := copyTabletStreams(tmpl)
+		for _, vrs := range tabletStreams {
+			for _, rule := range vrs.bls.Filter.Rules {
+				buf := &strings.Builder{}
+				t := template.Must(template.New("").Parse(rule.Filter))
+				err := t.Execute(buf, target.si.ShardName())
+				if err != nil {
+					return err
+				}
+				rule.Filter = buf.String()
+			}
+		}
+
+		buf := &strings.Builder{}
+		buf.WriteString("insert into _vt.vreplication(workflow, source, pos, max_tps, max_replication_lag, time_updated, transaction_timestamp, state, db_name) values ")
+		prefix := ""
+		for _, vrs := range tabletStreams {
+			fmt.Fprintf(buf, "%s(%v, %v, %v, %v, %v, %v, 0, '%v', %v)",
+				prefix,
+				encodeString(vrs.workflow),
+				encodeString(vrs.bls.String()),
+				encodeString(mysql.EncodePosition(vrs.pos)),
+				throttler.MaxRateModuleDisabled,
+				throttler.ReplicationLagModuleDisabled,
+				time.Now().Unix(),
+				binlogplayer.BlpStopped,
+				encodeString(target.master.DbName()))
+			prefix = ", "
+		}
+		_, err := sm.mi.wr.VReplicationExec(ctx, target.master.Alias, buf.String())
+		return err
+	})
+}
+
+func (sm *streamMigrater) cancelMigration(ctx context.Context) {
+	if sm.mi.migrationType == binlogdatapb.MigrationType_TABLES {
+		return
+	}
+	tabletStreams, err := sm.readSourceStreamsForCancel(ctx)
+	if err != nil {
+		sm.mi.wr.Logger().Errorf("Cancel migration failed: could not read streams metadata: %v", err)
+		return
+	}
+
+	// Ignore error. We still want to restart the source streams if deleteTargetStreans fails.
+	_ = sm.deleteTargetStreams(ctx, tabletStreams)
+
+	workflows := tabletStreamWorkflows(tabletStreams)
+	if len(workflows) == 0 {
+		return
+	}
+	workflowList := stringListify(workflows)
+	err = sm.mi.forAllSources(func(source *miSource) error {
+		query := fmt.Sprintf("update _vt.vreplication set state='Running', stop_pos=null, message='' where db_name=%s and workflow in (%s)", encodeString(source.master.DbName()), workflowList)
+		_, err := sm.mi.wr.VReplicationExec(ctx, source.master.Alias, query)
+		return err
+	})
+	if err != nil {
+		sm.mi.wr.Logger().Errorf("Cancel migration failed: could not restart source streams: %v", err)
+	}
+}
+
+func (sm *streamMigrater) deleteTargetStreams(ctx context.Context, tabletStreams []*vrStream) error {
+	workflows := tabletStreamWorkflows(tabletStreams)
+	if len(workflows) == 0 {
+		return nil
+	}
+	workflowList := stringListify(workflows)
+	err := sm.mi.forAllTargets(func(target *miTarget) error {
+		query := fmt.Sprintf("delete from _vt.vreplication where db_name=%s and workflow in (%s)", encodeString(target.master.DbName()), workflowList)
+		_, err := sm.mi.wr.VReplicationExec(ctx, target.master.Alias, query)
+		return err
+	})
+	if err != nil {
+		sm.mi.wr.Logger().Warningf("Could not delete migrated streams: %v", err)
```

Comment:
migrated streams? or target streams?

Reply:
target streams denote the ones that are doing the resharding. In this case, we're talking about the streams that we are migrating from the source to the target.